### PR TITLE
fix(service::repair): accept same nodePort with different protocols

### LINKED ISSUE #91404: API-server repair considers ports with same port number but different protocols as same
**What happened**:
I created several Workloads which require up to ten ports for both UDP and TCP protocol. Now my logs get spammed with `the node port [PORT] for service [SERVER] was assigned to multiple services; please recreate`.

**What you expected to happen**:
[repair.go::collectServiceNodePorts](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/portallocator/controller/repair.go#L203) should consider `PORT-PROTO`-combination instead of only `PORT`.

**How to reproduce it (as minimally and precisely as possible)**:
Deploy any service with a node-port with both protocols, e.g.:
```
apiVersion: v1
kind: Service
metadata:
  name: test
spec:
  ports:
  - port: 30080
    targetPort: 80
    protocol: UDP
    name: dummy
  - port: 30080
    targetPort: 80
    protocol: TCP
    name: http
```

**Anything else we need to know?**:
This was already acknowledged by @thockin here: https://github.com/kubernetes/kubernetes/issues/59119#issuecomment-490760644
However, as this thread addresses a different bug and seems kind of stuck, i'd like to suggest that this might be addressed in a new context.

**Environment**:
- Kubernetes version (use `kubectl version`): `v1.17.5`
- Cloud provider or hardware configuration: `Hetzner Cloud`
- OS (e.g: `cat /etc/os-release`): `Debian GNU/Linux 10 (buster)`
- Kernel (e.g. `uname -a`): `Linux server 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1+deb10u1 (2020-04-27) x86_64 GNU/Linux`
- Install tools: Deployed via Rancher v2.4.3

### PR DESCRIPTION
**Pull request by knight42**
Signed-off-by: knight42 <anonymousknight96@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #91404

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/registry/core/service/portallocator/controller/repair.go [hunk 1]
```diff
@@ -139,7 +140,7 @@ func (c *Repair) runOnce() error {
 			case nil:
 				if stored.Has(port) {
 					// remove it from the old set, so we can find leaks
-					stored.Release(port)
+					_ = stored.Release(port)
```
Comment (thockin) [inline pkg/registry/core/service/portallocator/controller/repair.go:143]: This is unnecessary

File: pkg/registry/core/service/portallocator/controller/repair_test.go [hunk 2]
```diff
@@ -203,3 +205,79 @@ func TestRepairWithExisting(t *testing.T) {
...
+			name: "duplicated nodePort with different protocol",
+			serviceSpec: corev1.ServiceSpec{
+				Ports: []corev1.ServicePort{
+					{NodePort: 111, Protocol: corev1.ProtocolTCP},
+					{NodePort: 112, Protocol: corev1.ProtocolTCP},
+					{NodePort: 111, Protocol: corev1.ProtocolUDP},
+				},
+			},
+			expected: []int{111, 112},
+		},
```
Comment (thockin) [inline pkg/registry/core/service/portallocator/controller/repair_test.go:236]: how about the case of duplicated port with same protocol

**Review by thockin — COMMENTED** [review]
Two really small comments and this is good to me.  Thanks!

/ok-to-test

File: pkg/registry/core/service/portallocator/controller/repair.go (hunk 1 above)
Comment (knight42) [reply pkg/registry/core/service/portallocator/controller/repair.go]: oh this is modified to make goland happy. restored.

File: pkg/registry/core/service/portallocator/controller/repair_test.go (hunk 2 above)
Comment (knight42) [reply pkg/registry/core/service/portallocator/controller/repair_test.go:258]: added.

**Comment by knight42** [comment]
@thockin hi! all tests passed now, could you `/approve` this pr?

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by bdowling** [comment]
Just fwiw, I came here thinking this was still an issue based on my observation.

But what I realized is that an "Update/Apply" to an existing service will only change 1 of the nodePorts that is on the same port.

If you delete and recreate the service, the duplicate TCP/UDP service accepts the same nodePort.


Server Version: v1.32.1

**Pull request merged**